### ISSUE TITLE
mod(pi,pi) not defined

### ISSUE THREAD

**Issue by andreasvarga** [JuliaLang/julia#56004]
The following shows the issue: 

```
julia> mod(float(pi),pi)
0.0

julia> mod(pi,float(pi))
0.0

julia> mod(pi,pi)
ERROR: mod not defined for Irrational{:π}
Stacktrace:
 [1] error(::String, ::String, ::Type)
   @ Base .\error.jl:44
 [2] no_op_err(name::String, T::Type)
   @ Base .\promotion.jl:506
 [3] mod(x::Irrational{:π}, y::Irrational{:π})
   @ Base .\promotion.jl:526
 [4] top-level scope
   @ REPL[109]:1
```
I wonder if this is intentional.

**Comment by ararslan** [comment]
I wouldn't say it's intentional necessarily so much as just a natural consequence of how things are set up. When the two arguments have `Real` but heterogeneous types, `promote` is called first, and the promotion rules for irrationals are such that they'll become floats upon `promote` with a different `Real` type. That's why your first two examples work. No promotion is necessary when the two arguments have the same type, but here we then hit the generic fallback method that just emits the error in your third example, since there is no applicable, more specific method defined.

Let's say we were to define such a method. What should it return? Generally speaking, for `T<:Real`, `mod(::T, ::T)` will return a `T`. (However, I don't believe we guarantee this.) Note that `typeof(π)` is `Irrational{:π}`. But `mod` of a number and itself should always be 0, and there is no 0 with type `Irrational{:π}`. So no matter what, `mod(::Irrational{:π}, ::Irrational{:π})` would have to return something that isn't an `Irrational{:π}`. One option is to return `zero(π)`, which is `false`. There is precedent for having irrationals convert to `Float64`s to facilitate some operations, so another option would be to do so here as well and return 0.0.

**Comment by andreasvarga** [comment]
Is this the same issue when we compute 

```
julia> pi-pi
0.0
```
?

**Comment by nsajko** [comment]
For exact arithmetic with irrational values, your best bet may be my package [TypeDomainNaturalNumbers.jl](https://gitlab.com/nsajko/TypeDomainNaturalNumbers.jl). That said, `mod(pi, pi)` isn't implemented yet, this is the ticket: [#54](https://gitlab.com/nsajko/TypeDomainNaturalNumbers.jl/-/issues/54).

**Comment by ararslan** [comment]
> Is this the same issue when we compute [...]

That's a case where we explicitly convert the irrationals to floats before doing the operation—it's actually a perfect example of the precedent I mentioned for doing that. 😄

Your example got me thinking: The docstring for `mod` describes the operation as being equivalent to `x - y * fld(x, y)` if computed without intermediate rounding, and if we take `x = y = π`, this is `π - π * floor(π / π)` (since there's no `fld(::Irrational{:π}, ::Irrational{:π})` either), which results in 0.0. This suggests that 0.0 would be a reasonable, consistent answer for `mod(π, π)`.

**Comment by andreasvarga** [comment]
I think this is the right approach. Since fully supporting operations with `Irrational` number seems to be quite a complex matter and is mostly irrelevant for numerical computations with floating point numbers, a pragmatic way to handle `Irrational` numbers in standard operations is to convert them to floats at each operation (as you described). In this vein, I would suggest to reflect on redefining `zero` and `one` for `Irrational` as:

```
zero(Irrational) = 0.0 
one(Irrational) = 1.0
```
instead of
```
zero(Irrational) = false
one(Irrational) = true
```

**Comment by nsajko** [comment]
No. The reason `Bool` was chosen is that it's the *weakest* type in terms of promotion. It may promote to any number.

**Comment by andreasvarga** [comment]
I think `rem` should be also addressed, because

```
julia> rem(pi,pi)
ERROR: rem not defined for Irrational{:π}
Stacktrace:
 [1] error(::String, ::String, ::Type)
   @ Base .\error.jl:44
 [2] no_op_err(name::String, T::Type)
   @ Base .\promotion.jl:506
 [3] rem(x::Irrational{:π}, y::Irrational{:π})
   @ Base .\promotion.jl:525
 [4] top-level scope
   @ REPL[32]:1

```
but 
```
julia> rem(pi,float(pi))
0.0

julia> rem(float(pi),pi)
0.0

```
